### ISSUE TITLE
Assignment2.py

### ISSUE THREAD

**Issue by realhsin** [realhsin/cmsl#1]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import IPython.display as disp

# Sigmoid function for activation
def sigmoid_function(x):
    return 1 / (1 + np.exp(-x))

# Perceptron training function for OR operation
def train_perceptron_or(X, y, learning_rate=0.1, epochs=1000):
    num_features = X.shape[1]
    weights = np.array([0.2, 0.1, -0.6])  # Initial weights
    bias = 0.1  # Initial bias

    training_history = []  # Store training history
    rms_per_epoch = []  # Store RMS values per epoch for graphing

    for epoch in range(epochs):
        total_error = 0

        for i in range(len(X)):
            # Compute weighted sum
            weighted_sum = np.dot(X[i], weights) + bias
            y_pred = sigmoid_function(weighted_sum)

            # Compute error
            error = y[i] - y_pred
            total_error += error ** 2

            # Update weights and bias
            weights += learning_rate * error * X[i]
            bias += learning_rate * error

            # Compute RMS error for current instance
            rms_error = np.sqrt(total_error / (i + 1))

            # Store training data history
            training_history.append([X[i][0], X[i][1], X[i][2], 
                                     weights[0], weights[1], weights[2], bias, 
                                     weighted_sum, y_pred, y[i], error, rms_error])

        # Compute final RMS error for the epoch
        final_rms_error = np.sqrt(total_error / len(X))
        rms_per_epoch.append(final_rms_error)  # Store for graphing

        training_history.append(["Epoch", epoch + 1, "", "", "", "", "", "", "", "", "", final_rms_error])

        # Stop training if RMS error is 0
        if final_rms_error == 0:
            break

    return weights, bias, training_history, rms_per_epoch  # Return RMS values too

# Prediction function to test the model
def predict(X_test, weights, bias):
    test_history = []
    total_error = 0

    for i in range(len(X_test)):
        weighted_sum = np.dot(X_test[i], weights) + bias
        y_pred = sigmoid_function(weighted_sum)
        predicted_label = 1 if y_pred >= 0.5 else 0  # Apply threshold

        # Compute error
        error = abs(predicted_label - y_pred)
        total_error += error ** 2

        # Compute RMS error for test data
        rms_error = np.sqrt(total_error / (i + 1))

        test_history.append([X_test[i][0], X_test[i][1], X_test[i][2], 
                             weights[0], weights[1], weights[2], bias, 
                             weighted_sum, y_pred, predicted_label, error, rms_error])

    return test_history

# ------------------------ TRAINING ------------------------ #
# Input features: [eye radius (ER), tail length (TL), tail width (TW)]
X_train = np.array([
    [4, 3, 1],    # Wolf
    [2, 3, 0.5],  # Dog
    [5, 4, 1.2],  # Wolf
    [4.5, 4.5, 1.1], # Wolf
    [1, 2.7, 0.5],  # Dog
    [1.7, 2.7, 0.7], # Dog
    [3.5, 3.5, 1.2], # Wolf
    [2.1, 2.3, 0.5], # Dog
    [2.3, 2.1, 0.7], # Dog
    [3.5, 3.2, 1.3]  # Wolf
])

# Expected output (labels): 1 = wolf, 0 = dog
y_train = np.array([1, 0, 1, 1, 0, 0, 1, 0, 0, 1])

# Train the perceptron
final_weights, final_bias, training_history, rms_per_epoch = train_perceptron_or(X_train, y_train)

# Convert training results to DataFrame
train_columns = ["X1", "X2", "X3", "W1", "W2", "W3", "Bias", "Sum", "Acti", "Target", "Error", "RMS"]
df_train = pd.DataFrame(training_history, columns=train_columns)

# Display training results
print("\nTraining Results:")
disp.display(df_train)

# ------------------------ TESTING ------------------------ #
# Given test data points
X_test = np.array([
    [5, 5.5, 2],   # Animal_Type_1
    [1.5, 1.5, 1.5] # Animal_Type_2
])

# Predict using the trained perceptron
test_results = predict(X_test, final_weights, final_bias)

# Convert test results to DataFrame
test_columns = ["X1", "X2", "X3", "W1", "W2", "W3", "Bias", "Sum", "Acti", "Predicted", "Error", "RMS"]
df_test = pd.DataFrame(test_results, columns=test_columns)

# Display test results
print("\nTesting Results:")
disp.display(df_test)

# ------------------------ PLOTTING RMS ERROR ------------------------ #
# Plot the learning curve
plt.figure(figsize=(10, 5))
plt.plot(range(1, len(rms_per_epoch) + 1), rms_per_epoch, marker='o', linestyle='-', color='b', label="RMS Error")
plt.xlabel("Epochs")
plt.ylabel("RMS Error")
plt.title("RMS Error vs. Epochs (Learning Curve)")
plt.legend()
plt.grid(True)
plt.show()
